src/hotspot/share/gc/shared/collectedHeap.cpp

```diff
@@ -603,6 +609,24 @@ void CollectedHeap::post_initialize() {
   initialize_serviceability();
 }
 
+bool CollectedHeap::is_shutting_down() const {
+  return Universe::is_shutting_down();
+}
+
+void CollectedHeap::stall_for_vm_shutdown() {
+  assert(is_shutting_down(), "Precondition");
+  // We use a timed wait (2 seconds) instead of an indefinite wait to avoid deadlock
+  // if the VM shutdown triggers a GC.
+  // The 2-second timeout is:
+  //   - long enough to keep daemon threads stalled, while the shutdown
+  //     sequence completes in the common case.
+  //   - short enough to avoid excessive stall time if the shutdown itself
+  //     triggers a GC.
+  MonitorLocker ml(VMExit_lock);
+  ml.wait(2 * MILLIUNITS);
```

Comment:
I think one can use `ThreadBlockInVM` + sleep to achieve the blocking-current-thread purpose. Then, there is no need for a new lock, as there is no critical-region anyway.

Reply:
Changed to just using sleep.